src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_TupleBinaryOperator.cs

```diff
@@ -0,0 +1,355 @@
+                if (operatorKind == BinaryOperatorKind.Equal)
+                {
+                    return _factory.Not(MakeUnaryOperator(UnaryOperatorKind.DynamicFalse, left.Syntax, method: null, dynamicResult, boolType));
+                }
+                else
+                {
+                    return MakeUnaryOperator(UnaryOperatorKind.DynamicTrue, left.Syntax, method: null, dynamicResult, boolType);
+                }
+            }
+
+            if (left.IsLiteralNull() && right.IsLiteralNull())
+            {
+                // For `null == null` this is special-cased during initial binding
+                return new BoundLiteral(left.Syntax, ConstantValue.Create(operatorKind == BinaryOperatorKind.Equal), boolType);
+            }
+
+            // PROTOTYPE(tuple-equality) checked
+            // We leave the null literal unconverted because MakeBinaryOperator has special rules for it
+            BoundExpression convertedLeft = left.IsLiteralNull() ? left : MakeConversionNode(left.Syntax, left, single.LeftConversion, single.LeftConvertedType, @checked: false);
+            BoundExpression convertedRight = right.IsLiteralNull() ? right : MakeConversionNode(right.Syntax, right, single.RightConversion, single.RightConvertedType, @checked: false);
+
+            BoundExpression binary = MakeBinaryOperator(_factory.Syntax, single.Kind, convertedLeft, convertedRight, single.MethodSymbolOpt?.ReturnType ?? boolType, single.MethodSymbolOpt);
+            UnaryOperatorSignature boolOperator = single.BoolOperator;
+            Conversion boolConversion = single.ConversionForBool;
+
+            BoundExpression result;
+            if (boolOperator.Kind != UnaryOperatorKind.Error)
+            {
+                // Produce
+                // !((left == right).op_false)
+                // (left != right).op_true
+                BoundExpression convertedBinary = MakeConversionNode(_factory.Syntax, binary, boolConversion, boolOperator.OperandType, @checked: false);
+
+                Debug.Assert(boolOperator.ReturnType.SpecialType == SpecialType.System_Boolean);
+                result = MakeUnaryOperator(boolOperator.Kind, binary.Syntax, boolOperator.Method, convertedBinary, boolType);
+
+                if (operatorKind == BinaryOperatorKind.Equal)
+                {
+                    result = _factory.Not(result);
+                }
+            }
+            else if (boolConversion != Conversion.Identity)
+            {
+                // Produce
+                // (bool)(left == right)
+                // (bool)(left != right)
+                result = MakeConversionNode(_factory.Syntax, binary, boolConversion, boolType, @checked: false);
+            }
+            else
+            {
+                result = binary;
+            }
+
+            return result;
+        }
+
+        private class TupleOperatorSideEffectsAndTemps
```

Comment:
>TupleOperatorSideEffectsAndTemps [](start = 22, length = 32)

With the current approach, it feels like this type does not give us any advantage. Side-effects are only collected locally in VisitTupleBinaryOperator and, I think, the code would be much cleaner if other helpers would keep track of their temps and create sequences accordingly. Even, we don't change temp management, it is still enough to pass a the same builder for side-effects to ReplaceTerminalElementsWithTemps, and simply pass a builder for temps to other helpers.